ledger/participant-integration-api/src/main/scala/platform/store/backend/common/EventStorageBackendTemplate.scala

```diff
@@ -955,19 +929,123 @@ abstract class EventStorageBackendTemplate(
-      UNION ALL
-      (${selectFrom("participant_events_create")})
-      UNION ALL
-      (${selectFrom("participant_events_non_consuming_exercise")})
-    ) participant_events"""
-      .as(get[Long](1).?.single)(connection)
+    SQL"""
+         SELECT
+            event_sequential_id_last
+         FROM
+            participant_transaction_meta
+         WHERE
+            event_offset = (SELECT MAX(event_offset) FROM participant_transaction_meta WHERE event_offset <= $offset)
+       """.as(get[Long](1).singleOpt)(connection)
+  }
+
+  private def pruneIdFilterCreateStakeholder(pruneUpToInclusive: Offset): SimpleSql[Row] =
+    pruneIdFilterCreate(
+      tableName = "pe_create_id_filter_stakeholder",
+      pruneUpToInclusive = pruneUpToInclusive,
+    )
+
+  private def pruneIdFilterCreateNonStakeholderInformee(
+      pruneUpToInclusive: Offset
+  ): SimpleSql[Row] =
+    pruneIdFilterCreate(
+      tableName = "pe_create_id_filter_non_stakeholder_informee",
+      pruneUpToInclusive = pruneUpToInclusive,
+    )
+
+  private def pruneIdFilterConsumingStakeholder(pruneUpToInclusive: Offset): SimpleSql[Row] =
+    pruneIdFilterConsumingOrNonConsuming(
+      idFilterTableName = "pe_consuming_id_filter_stakeholder",
+      eventsTableName = "participant_events_consuming_exercise",
+      pruneUpToInclusive = pruneUpToInclusive,
+    )
+
+  private def pruneIdFilterConsumingNonStakeholderInformee(
+      pruneUpToInclusive: Offset
+  ): SimpleSql[Row] = {
+    pruneIdFilterConsumingOrNonConsuming(
+      idFilterTableName = "pe_consuming_id_filter_non_stakeholder_informee",
+      eventsTableName = "participant_events_consuming_exercise",
+      pruneUpToInclusive = pruneUpToInclusive,
+    )
+  }
+
+  private def pruneIdFilterNonConsumingInformee(pruneUpToInclusive: Offset): SimpleSql[Row] =
+    pruneIdFilterConsumingOrNonConsuming(
+      idFilterTableName = "pe_non_consuming_id_filter_informee",
+      eventsTableName = "participant_events_non_consuming_exercise",
+      pruneUpToInclusive = pruneUpToInclusive,
+    )
+
+  /** Callers can call it only once pruning of create, consuming and non-consuming event tables has already finished.
+    * The implementation assumes that these tables have already been pruned.
+    */
+  private def pruneTransactionMeta(pruneUpToInclusive: Offset): SimpleSql[Row] = {
+    import com.daml.platform.store.backend.Conversions.OffsetToStatement
+    SQL"""
+         DELETE FROM
+            participant_transaction_meta m
+         WHERE
+          m.event_offset <= $pruneUpToInclusive
+          AND
+          NOT EXISTS (
+            SELECT 1 FROM participant_events_create c
+            WHERE
+              c.event_sequential_id >= m.event_sequential_id_first
+              AND
+              c.event_sequential_id <= m.event_sequential_id_last
+          )
+       """
+  }
+
+  // TODO etq: Currently we query two additional tables: create and consuming events tables.
```

Comment:
I don't understand, as it stands this query will stop working if it's called after create and archive events have been pruned, so this implies that there is already ordering.
Given the complexity of this query I would revert the ordering and demand that the create/archive tables have been pruned.

Reply:
This query implementation and the implied ordering (that an id filter table needs to be pruned before an events table) is preceding ETQ and was introduced during ACS work.
I opted to preserve the old philosophy and handle new id filter tables analogously to limit the risk of unforeseen consquences.

My preference would be to simplify those queries and swap the ordering in a separate PR that is only focusing on that aspect